src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,430 @@
+            {
+                if (RequiredDelimiterIsMissing(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (StatementClosingDelimiterIsMissing(currentNode))
+            {
+                // Example: missing final `)` in `do { } while (x$$`
+                return;
+            }
+
+            var semicolonPosition = GetSemicolonLocation(root, currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(semicolonPosition));
+        }
+
+        private int GetSemicolonLocation(SyntaxNode root, SyntaxNode currentNode, int caretPosition)
+        {
+            int end;
+            if (currentNode.IsKind(SyntaxKind.ForStatement))
+            {
+                // in for statements, semicolon can go after initializer or after condition, depending on where the caret is located
+                var forStatementSyntax = (ForStatementSyntax)currentNode;
+                if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Condition.FullSpan.End;
+                }
+                else if (CaretIsInForStatementDeclaration(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Declaration.FullSpan.End;
+                }
+                else if (CaretIsInForStatementInitializers(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Initializers.FullSpan.End;
+                }
+                else
+                {
+                    // Should not be reachable because we returned earlier for this case
+                    end = 0;
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+            else
+            {
+                end = currentNode.Span.End;
+            }
+
+            // If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.
+            // To avoid placing the semicolon on the next line or after comments, we need to adjust the position.
+            return root.FindToken(end).GetPreviousToken().Span.End;
+        }
+
+        private static bool CaretIsInForStatementCondition(int caretPosition, ForStatementSyntax forStatementSyntax)
+        {
+            return caretPosition > forStatementSyntax.Condition.SpanStart && caretPosition < forStatementSyntax.Condition.Span.End;
```

Comment:
`condition` can be null, but you don't seem to be checking for that case.

Reply:
I had one test for that, but this prompted me to think of others.  Code updated as well.